internal/services/manager.go

```diff
@@ -122,23 +287,108 @@ func StopProcessManager(
 	return nil
 }
 
-func ProcessManagerIsRunning() bool {
-	pid, err := os.ReadFile(processComposePidfile)
+func StopAllProcessManagers(ctx context.Context, w io.Writer) error {
+	configPath, err := globalProcessComposeJSONPath()
 	if err != nil {
-		return false
+		return fmt.Errorf("failed to get config path: %w", err)
 	}
 
-	pidToInt, _ := strconv.Atoi(string(pid))
-	process, err := os.FindProcess(pidToInt)
+	// Open the config file, defer closing to the end of the scope
+	globalConfigFile, err := os.OpenFile(configPath, os.O_RDWR|os.O_CREATE, 0666)
+	if err != nil {
+		return fmt.Errorf("failed to open config file: %w", err)
+	}
+	defer globalConfigFile.Close()
+
+	// Lock the config file, defer unlocking to the end of the scope
+
+	lockFile(globalConfigFile.Fd())
+	defer unlockFile(globalConfigFile.Fd())
+	config := readGlobalProcessComposeJSON(configPath)
+
+	for _, project := range config.Instances {
+		pid, _ := os.FindProcess(project.Pid)
+		err := pid.Signal(os.Interrupt)
+		if err != nil {
+			fmt.Printf("process-compose is not running. To start it, run `devbox services up`")
+		}
+	}
+
+	config.Instances = make(map[string]instance)
+
+	err = writeGlobalProcessComposeJSON(config)
+	if err != nil {
+		return fmt.Errorf("failed to write global process-compose config: %w", err)
+	}
+
+	return nil
+}
+
+func ProcessManagerIsRunning(projectDir string) bool {
+
+	configPath, err := globalProcessComposeJSONPath()
 	if err != nil {
-		fmt.Printf("Error: %s \n", err)
 		return false
 	}
 
+	config := readGlobalProcessComposeJSON(configPath)
+
+	project, ok := config.Instances[projectDir]
+	if !ok {
+		return false
+	}
+
+	process, _ := os.FindProcess(project.Pid)
+
 	err = process.Signal(syscall.Signal(0))
 	if err != nil {
 		fmt.Printf("Error: %s \n", err)
+		delete(config.Instances, projectDir)
+		_ = writeGlobalProcessComposeJSON(config)
 		return false
 	}
+
 	return true
 }
+
+func GetProcessManagerPort(projectDir string) (int, error) {
+	path, err := globalProcessComposeJSONPath()
+	if err != nil {
+		return 0, fmt.Errorf("process-compose is not running or it's config is missing. To start it, run `devbox services up`")
+	}
+
+	config := readGlobalProcessComposeJSON(path)
+
+	project, ok := config.Instances[projectDir]
+	if !ok {
+		return 0, fmt.Errorf("process-compose is not running or it's config is missing. To start it, run `devbox services up`")
+	}
+
+	return project.Port, nil
+}
+
+func lockFile(fd uintptr) {
+
+	lock := syscall.Flock_t{
+		Type:   syscall.F_WRLCK,
+		Whence: int16(io.SeekStart),
+		Start:  0,
+		Len:    0,
+	}
+
+	if err := syscall.FcntlFlock(fd, syscall.F_SETLK, &lock); err != nil {
+		fmt.Printf("Error acquiring lock: %s \n", err)
+	}
```

Comment:
why not return the error message and fail if we don't get the lock?

Continuing to write could lead to some weird errors because of clobbering writes.

Reply:
We now return an error from the lock